File: app/ios/Shared/LiveActivity/UnifiedEew.swift
import Foundation

struct UnifiedEewLocation: Codable, Hashable, Sendable {
    let regionName: String
    let forecastIntensity: IntensityValue?
    let forecastLpgmIntensity: LpgmIntensityValue?
    let arrivalTime: LiveActivityTimestamp?
    let isPlum: Bool?
    let isWarning: Bool?

    private enum CodingKeys: String, CodingKey {
        case regionName
        case forecastIntensity
        case forecastLpgmIntensity
        case arrivalTime
        case isPlum
        case isWarning
    }

    init(from decoder: Decoder) throws {
        let container = try decoder.container(keyedBy: CodingKeys.self)
        regionName = try container.decode(String.self, forKey: .regionName)
        forecastIntensity = try container.decodeOptionalNonNull(
            IntensityValue.self,
            forKey: .forecastIntensity
        )
        forecastLpgmIntensity = try container.decodeOptionalNonNull(
            LpgmIntensityValue.self,
            forKey: .forecastLpgmIntensity
        )
        arrivalTime = try container.decodeOptionalNonNull(
            LiveActivityTimestamp.self,
            forKey: .arrivalTime
        )
        isPlum = try container.decodeOptionalNonNull(Bool.self, forKey: .isPlum)
        isWarning = try container.decodeOptionalNonNull(Bool.self, forKey: .isWarning)
        if isPlum == false {
            throw DecodingError.dataCorruptedError(
                forKey: .isPlum,
                in: container,
                debugDescription: "isPlum may only be true when present"
            )
        }
    }
}

struct UnifiedEew: Codable, Hashable, Sendable {
    let eventId: String
    let headline: String
    let hypocenterName: String?
    let magnitude: Double?
    let depth: Double?
    let time: LiveActivityTimestamp?
    let isOriginTime: Bool
    let maxIntensity: IntensityValue?
    let serialNo: Int
    let isFinal: Bool
    let isWarning: Bool
    let isCanceled: Bool
    let isPlum: Bool
    let isLevel: Bool
    let isOnePoint: Bool
    let issuedAt: LiveActivityTimestamp
    let location: UnifiedEewLocation?

    private enum CodingKeys: String, CodingKey {
        case eventId
        case headline
        case hypocenterName
        case magnitude
        case depth
        case time
        case isOriginTime
        case maxIntensity
        case serialNo
        case isFinal
        case isWarning
        case isCanceled
        case isPlum
        case isLevel
        case isOnePoint
        case issuedAt
        case location
    }

    init(from decoder: Decoder) throws {
        let container = try decoder.container(keyedBy: CodingKeys.self)
        eventId = try container.decodeNonemptyString(forKey: .eventId)
        headline = try container.decode(String.self, forKey: .headline)
        hypocenterName = try container.decodeRequiredNullable(String.self, forKey: .hypocenterName)
        magnitude = try container.decodeRequiredNullableFiniteDouble(forKey: .magnitude)
        depth = try container.decodeRequiredNullableFiniteDouble(forKey: .depth)
        time = try container.decodeRequiredNullable(LiveActivityTimestamp.self, forKey: .time)
        isOriginTime = try container.decode(Bool.self, forKey: .isOriginTime)
        maxIntensity = try container.decodeRequiredNullable(IntensityValue.self, forKey: .maxIntensity)
        serialNo = try container.decode(Int.self, forKey: .serialNo)
        guard serialNo >= 0 else {
            throw DecodingError.dataCorruptedError(
                forKey: .serialNo,
                in: container,
                debugDescription: "serialNo must be nonnegative"
            )
        }
        isFinal = try container.decode(Bool.self, forKey: .isFinal)
        isWarning = try container.decode(Bool.self, forKey: .isWarning)
        isCanceled = try container.decode(Bool.self, forKey: .isCanceled)
        isPlum = try container.decode(Bool.self, forKey: .isPlum)
        isLevel = try container.decode(Bool.self, forKey: .isLevel)
        isOnePoint = try container.decode(Bool.self, forKey: .isOnePoint)
        issuedAt = try container.decode(LiveActivityTimestamp.self, forKey: .issuedAt)
        location = try container.decodeRequiredNullable(UnifiedEewLocation.self, forKey: .location)
    }

    func encode(to encoder: Encoder) throws {
        var container = encoder.container(keyedBy: CodingKeys.self)
        try container.encode(eventId, forKey: .eventId)
        try container.encode(headline, forKey: .headline)
        try container.encodeRequiredNullable(hypocenterName, forKey: .hypocenterName)
        try container.encodeRequiredNullable(magnitude, forKey: .magnitude)
        try container.encodeRequiredNullable(depth, forKey: .depth)
        try container.encodeRequiredNullable(time, forKey: .time)
        try container.encode(isOriginTime, forKey: .isOriginTime)
        try container.encodeRequiredNullable(maxIntensity, forKey: .maxIntensity)
        try container.encode(serialNo, forKey: .serialNo)
        try container.encode(isFinal, forKey: .isFinal)
        try container.encode(isWarning, forKey: .isWarning)
        try container.encode(isCanceled, forKey: .isCanceled)
        try container.encode(isPlum, forKey: .isPlum)
        try container.encode(isLevel, forKey: .isLevel)
        try container.encode(isOnePoint, forKey: .isOnePoint)
        try container.encode(issuedAt, forKey: .issuedAt)
        try container.encodeRequiredNullable(location, forKey: .location)
    }
}

File: app/lib/feature/live_activity/data/model/unified_eew.dart
// ignore_for_file: unnecessary_type_name_in_constructor

import 'package:eqmonitor/core/model/intensity/jma_intensity.dart';
import 'package:eqmonitor/core/model/intensity/jma_lpgm_intensity.dart';
import 'package:eqmonitor/feature/live_activity/data/model/unified_live_activity_json_converter.dart';
import 'package:freezed_annotation/freezed_annotation.dart';

part 'unified_eew.freezed.dart';

@freezed
abstract class UnifiedEewLocation with _$UnifiedEewLocation {
  const UnifiedEewLocation._();

  const factory UnifiedEewLocation({
    required String regionName,
    required JmaIntensity? forecastIntensity,
    required JmaLpgmIntensity? forecastLpgmIntensity,
    required DateTime? arrivalTime,
    required bool? isPlum,
    required bool? isWarning,
  }) = _UnifiedEewLocation;

  factory UnifiedEewLocation.fromJson(Map<String, dynamic> json) {
    final reader = UnifiedLiveActivityJsonReader(json, label: 'eew.location');
    final intensity = reader.optionalNonNullString('forecastIntensity');
    final lpgm = reader.optionalNonNullString('forecastLpgmIntensity');
    final isPlum = reader.optionalNonNullBool('isPlum');
    if (isPlum == false) {
      throw const FormatException('eew.location.isPlum は true のみ指定できます');
    }
    return UnifiedEewLocation(
      regionName: reader.requiredString('regionName'),
      forecastIntensity: intensity == null
          ? null
          : UnifiedLiveActivityWire.intensityFromJson(intensity),
      forecastLpgmIntensity: lpgm == null
          ? null
          : UnifiedLiveActivityWire.lpgmFromJson(lpgm),
      arrivalTime: reader.optionalNonNullDateTime('arrivalTime'),
      isPlum: isPlum,
      isWarning: reader.optionalNonNullBool('isWarning'),
    );
  }

  Map<String, dynamic> toJson() => <String, dynamic>{
    'regionName': regionName,
    if (forecastIntensity case final value?)
      'forecastIntensity': UnifiedLiveActivityWire.intensityToJson(value),
    if (forecastLpgmIntensity case final value?)
      'forecastLpgmIntensity': UnifiedLiveActivityWire.lpgmToJson(value),
    if (arrivalTime case final value?)
      'arrivalTime': UnifiedLiveActivityWire.dateTimeToJson(value),
    if (isPlum != null) 'isPlum': isPlum,
    if (isWarning != null) 'isWarning': isWarning,
  };
}

@freezed
abstract class UnifiedEew with _$UnifiedEew {
  const UnifiedEew._();

  const factory UnifiedEew({
    required String eventId,
    required String headline,
    required String? hypocenterName,
    required double? magnitude,
    required double? depth,
    required DateTime? time,
    required bool isOriginTime,
    required JmaIntensity? maxIntensity,
    required int serialNo,
    required bool isFinal,
    required bool isWarning,
    required bool isCanceled,
    required bool isPlum,
    required bool isLevel,
    required bool isOnePoint,
    required DateTime issuedAt,
    required UnifiedEewLocation? location,
  }) = _UnifiedEew;

  factory UnifiedEew.fromJson(Map<String, dynamic> json) {
    final reader = UnifiedLiveActivityJsonReader(json, label: 'eew');
    final maxIntensity = reader.requiredNullableString('maxIntensity');
    final location = reader.requiredNullableMap('location');
    return UnifiedEew(
      eventId: reader.requiredString('eventId', nonEmpty: true),
      headline: reader.requiredString('headline'),
      hypocenterName: reader.requiredNullableString('hypocenterName'),
      magnitude: reader.requiredNullableFiniteDouble('magnitude'),
      depth: reader.requiredNullableFiniteDouble('depth'),
      time: reader.requiredNullableDateTime('time'),
      isOriginTime: reader.requiredBool('isOriginTime'),
      maxIntensity: maxIntensity == null
          ? null
          : UnifiedLiveActivityWire.intensityFromJson(maxIntensity),
      serialNo: reader.requiredInt('serialNo', minimum: 0),
      isFinal: reader.requiredBool('isFinal'),
      isWarning: reader.requiredBool('isWarning'),
      isCanceled: reader.requiredBool('isCanceled'),
      isPlum: reader.requiredBool('isPlum'),
      isLevel: reader.requiredBool('isLevel'),
      isOnePoint: reader.requiredBool('isOnePoint'),
      issuedAt: reader.requiredDateTime('issuedAt'),
      location: location == null ? null : UnifiedEewLocation.fromJson(location),
    );
  }

  Map<String, dynamic> toJson() => <String, dynamic>{
    'eventId': eventId,
    'headline': headline,
    'hypocenterName': hypocenterName,
    'magnitude': magnitude,
    'depth': depth,
    'time': switch (time) {
      final value? => UnifiedLiveActivityWire.dateTimeToJson(value),
      null => null,
    },
    'isOriginTime': isOriginTime,
    'maxIntensity': switch (maxIntensity) {
      final value? => UnifiedLiveActivityWire.intensityToJson(value),
      null => null,
    },
    'serialNo': serialNo,
    'isFinal': isFinal,
    'isWarning': isWarning,
    'isCanceled': isCanceled,
    'isPlum': isPlum,
    'isLevel': isLevel,
    'isOnePoint': isOnePoint,
    'issuedAt': UnifiedLiveActivityWire.dateTimeToJson(issuedAt),
    'location': location?.toJson(),
  };
}
